module3/homework/passwordCheck/validation.cpp

```diff
@@ -1,2 +1,69 @@
+#include <cstring>
+#include <algorithm>
+
+
 #include "validation.hpp"
-// TODO: Put implementations here
\ No newline at end of file
+
+const int8_t minPasswordLen = 9;
+std::string getErrorMessage(ErrorCode code)
+{
+    switch(code)
+    {
+        case ErrorCode::Ok : {return "OK";}
+        case ErrorCode::PasswordNeedsAtLeastNineCharacters : 
+                    {return "ERROR. Password needs at least nine characters.";}
+        case ErrorCode::PasswordNeedsAtLeastOneNumber : 
+                    {return "ERROR. Password needs at least one number.";}
+        case ErrorCode::PasswordNeedsAtLeastOneSpecialCharacter : 
+                    {return "ERROR. Password needs at least one special character.";}
+        case ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter : 
+                    {return "ERROR. Password needs at least one uppercase letter.";}
+        case ErrorCode::PasswordsDoesNotMatch: 
+                    {return "ERROR. Password does not match.";}
+    }
+}
+
+bool doesPasswordsMatch(std::string password, std::string repeated_password)
+{
+    if(password != repeated_password) return false;
+    return true;
+}
+
+ErrorCode checkPasswordRules(std::string const password)
+{
+    if(password.length() < minPasswordLen) return ErrorCode::PasswordNeedsAtLeastNineCharacters;
+
+    if(std::find_if(password.begin(), 
+        password.end(), isdigit) == password.end())
+        return ErrorCode::PasswordNeedsAtLeastOneNumber;
+
+    if(std::find_if(password.begin(), 
+        password.end(), isupper) == password.end())
+        return ErrorCode::PasswordNeedsAtLeastOneUppercaseLetter;
+
+    bool passwordContainsSpecialCharacter = false;
+    std::string const listOfTheSpecialCharacters = "!@#$%^&*";
+    for(const char& ch : listOfTheSpecialCharacters)
+    {
+        if(std::any_of(password.begin(), password.end(),
+                [ch](const char& l){return l == ch;}))
+                {
+                    passwordContainsSpecialCharacter = true;
+                    break;
```

Comment:
I really don't like this formatting. It's hiding the intention of this code. For 3 minutes I was thinking that all this code is a part of lambda. Please use code autoformatting in your IDE.
```suggestion
        if(std::any_of(password.begin(), password.end(), [ch](const char& l){return l == ch;}))
        {
            passwordContainsSpecialCharacter = true;
            break;
        }
```